src/libDirectoryService/FinalBlockPreProcessing.cpp

```diff
@@ -290,17 +328,63 @@ bool DirectoryService::RunConsensusOnFinalBlockWhenDSPrimary() {
         << "] BEGIN");
   }
 
-  auto announcementGeneratorFunc =
+  auto preprepAnnouncementGeneratorFunc =
       [this](bytes& dst, unsigned int offset, const uint32_t consensusID,
              const uint64_t blockNumber, const bytes& blockHash,
              const uint16_t leaderID, const PairOfKey& leaderKey,
              bytes& messageToCosign) mutable -> bool {
     return Messenger::SetDSFinalBlockAnnouncement(
         dst, offset, consensusID, blockNumber, blockHash, leaderID, leaderKey,
-        *m_finalBlock, m_mediator.m_node->m_microblock, messageToCosign);
+        *m_finalBlock, m_mediator.m_node->m_prePrepMicroblock, messageToCosign);
   };
 
-  cl->StartConsensus(announcementGeneratorFunc, BROADCAST_GOSSIP_MODE);
+  auto newFBAnnouncementReadinessFunc =
+      [this](bytes& newAnnouncement, unsigned int offset,
+             const uint32_t consensusID, const uint64_t blockNumber,
+             const bytes& blockHash, const uint16_t leaderID,
+             const PairOfKey& leaderKey,
+             bytes& messageToCosign) mutable -> bool {
+    // wait for complete final block (with complete microblock) being ready by
+    // me (leader)
+    if (!WaitUntilCompleteFinalBlockIsReady()) {
+      return false;
+    }
+
+    return Messenger::SetDSFinalBlockAnnouncement(
+        newAnnouncement, offset, consensusID, blockNumber, blockHash, leaderID,
+        leaderKey, *m_finalBlock, m_mediator.m_node->m_microblock,
+        messageToCosign);
+  };
+
+  cl->StartConsensus(preprepAnnouncementGeneratorFunc,
+                     newFBAnnouncementReadinessFunc, BROADCAST_GOSSIP_MODE);
+
+  SetState(FINALBLOCK_CONSENSUS);
+
+  if (m_mediator.ToProcessTransaction()) {
+    m_mediator.m_node->ProcessTransactionWhenShardLeader(m_microBlockGasLimit);
+    if (!AccountStore::GetInstance().SerializeDelta()) {
+      LOG_GENERAL(WARNING, "AccountStore::SerializeDelta failed");
+      return false;
+    }
+  }
+  AccountStore::GetInstance().CommitTempRevertible();
+
+  if (!m_mediator.m_node->ComposeMicroBlock(m_microBlockGasLimit)) {
+    LOG_GENERAL(WARNING, "DS ComposeMicroBlock Failed");
+    m_mediator.m_node->m_microblock = nullptr;
+  } else {
+    lock_guard<mutex> g(m_mutexMicroBlocks);
+    m_microBlocks[m_mediator.m_currentEpochNum].emplace(
+        *(m_mediator.m_node->m_microblock));
+  }
+
+  // stores it in m_finalBlock
+  if (!ComposeFinalBlock()) {
+    LOG_EPOCH(WARNING, m_mediator.m_currentEpochNum,
```

Comment:
Might wanna change the message to `DS ComposeFinalBlock (prep) Failed` for consistency with `ComposePrePrepMicroBlock` above and `ComposeMicroBlock` below.